### PR TITLE
Implement rm in ramalama

### PR DESCRIPTION
**Pull request by rhatdan**
We must delete the symlink and garbage collect any model that does not
have an associated symlink.

We do garbage collection of everything, because some old sha's etc. may
no longer have symlink references due to updates.

Co-Authored-By: Daniel J Walsh <dwalsh@redhat.com>
Signed-off-by: Eric Curtin <ecurtin@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@ericcurtin this is currently broken because of the way we are storing ollama models.  Why aren't we storing like we do huggingface models?

**Comment by ericcurtin** [comment]
@rhatdan

for Ollama support, we wrote our own client basically, using various curl commands (http requests).

for huggingface we use the official huggingface client:

huggingface_hub[cli]==0.24.2

for oci we use the omlmd client:

omlmd==0.1.4

These clients store the .gguf's in their respective directories here (but each client does it differently):

~/.local/share/ramalama/repos/

This dir stores the models in a more structured way (as symlinks):

~/.local/share/ramalama/models/

this is all open to change... The clients for each protocol are fundamentally different right now...

**Comment by ericcurtin** [comment]
But I see what you mean...

Maybe we should store like huggingface...

That sha256: name is really useful though (that's the filename in Ollama), you can validate if the file is fully downloaded and correct.

**Comment by ericcurtin** [comment]
omlmd could be completely different again, haven't tried it yet.

**Comment by ericcurtin** [comment]
Huggingface client seems to create directories like this with some metadata:

.huggingface
.cache/huggingface

so maybe they don't need the sha256:

**Comment by ericcurtin** [comment]
> huggingface-cli delete-cache

can also delete GB's of data

**Comment by ericcurtin** [comment]
I completely forgot a super important reason for the sha256: filename being important... De-duplication, you will never download duplicate files this way as you will notice the checksum exists already

**Comment by rhatdan** [comment]
@ericcurtin Since you understand how this is supposed to be, do you want to take over this PR?

**Comment by ericcurtin** [comment]
Sure I'll get to it tomorrow

**Comment by ericcurtin** [comment]
I will get to this Monday @rhatdan haven't forgotten...

If it unblocks you we could just rm the symlink for now and do a follow on PR that actually rm's the real file when appropriate

**Comment by ericcurtin** [comment]
I rebased this and fixed some conflicts

**Comment by ericcurtin** [comment]
I have no idea right now why this is failing CI... Will try and figure out tomorrow, seems fine locally

**Comment by rhatdan** [comment]
The mac_steps and linux_steps part of the ci.sh are screwing up the system, those should not be run within tests.

https://paste.centos.org/view/af6ab85c

Here are the other changes required. ^^

File: ramalama/cli.py [hunk 1]
```diff
@@ -256,6 +257,17 @@ def version_parser(subparsers):
     parser.set_defaults(func=version_cli)
 
 
+def rm_parser(subparsers):
+    parser = subparsers.add_parser("rm", help="Remove AI model from local storage")
+    parser.add_argument("model", help="Name of the model to remove")
```
Comment (rhatdan) [inline ramalama/cli.py:262]: Capitalize model

File: ramalama/cli.py [hunk 2]
```diff
@@ -256,6 +257,17 @@ def version_parser(subparsers):
...
+    parser.set_defaults(func=rm_cli)
+
+
+def rm_cli(args):
+    model = New(args.model)
```
Comment (rhatdan) [inline ramalama/cli.py:267]: Capitalize model

File: test/ci.sh [hunk 3]
```diff
@@ -40,6 +40,12 @@ tests() {
   ./${binfile} ls | grep tiny-vicuna-1b
   ./${binfile} ls | grep NAME
   ./${binfile} ls | grep oci://quay.io/mmortari/gguf-py-example/v1/example.gguf
+  ./${binfile} rm ollama://ben1t0/tiny-llm:latest
```
Comment (rhatdan) [inline test/ci.sh:43]: Should add this as a test to bats.

File: test/ci.sh [hunk 4]
```diff
@@ -72,13 +78,13 @@ main() {
     linux_steps
   fi
 
+  $maybe_sudo rm -rf /usr/share/ramalama /opt/homebrew/share/ramalama /usr/local/share/ramalama
```
Comment (rhatdan) [inline test/ci.sh:81]: The problem with this is it modifies the host system when users run `make test`
 If we want to test this, we should just do it in the ci/cd system not as part of the `make test` job.

Comment (ericcurtin) [reply test/ci.sh:81]: We have to at least run:

mac_steps -> ./install.py

or some equivalent... To install the python dependancies for macOS...

File: ramalama/cli.py (hunk 2 above)
Comment (ericcurtin) [reply ramalama/cli.py:267]: There's a lot of similar references, do I capitalise everywhere? :

```
ramalama/cli.py
80:    model = New(str(transport))
97:    model = New(str(transport))
210:    model = New(args.model)
226:    model = New(args.model)
239:    model = New(args.model)
251:    model = New(args.model)
267:    model = New(args.model)

ramalama/model.py
9:    model = ""
15:        self.model = model
```

Comment (ericcurtin) [reply ramalama/cli.py:267]: I think I see what you mean now, I'll try to be consistent with your recent PR

**Comment by rhatdan** [comment]
Ok lets get this merged, and then I can start turning on tests and probably add some more options.

**Comment by rhatdan** [comment]
LGTM

File: ramalama/cli.py (hunk 1 above)
Comment (ericcurtin) [reply ramalama/cli.py]: Done

File: ramalama/cli.py (hunk 2 above)
Comment (ericcurtin) [reply ramalama/cli.py]: Done

**Comment by ericcurtin** [comment]
I'm fighting with builds at the moment, it might be green this one

**Pull request merged**